core/primitives/src/errors.rs

```diff
@@ -84,6 +92,9 @@ pub enum StorageError {
     /// panic in every place that produces this error.
     /// We can check if db is corrupted by verifying everything in the state trie.
     StorageInconsistentState(String),
+    /// Error thrown when a given value cannot be parsed as a TrieNode.
+    /// This probably means that this is a 'value' field (as we store both in the same colState).
+    NotANode(NotANodeError),
```

Comment:
`NotANode` is a kind of `StorageInconsistentState`. If you want to have it well-typed, then I sugest adding

```
enum StorageInconsistentState { Other(String), NotANodeError(NotANodeError) }
```

Reply:
it is hard to say if this is the 'inconsistency'...
This is one weakness of our design - where we put two types of objects into one column (Nodes and values), without any indication on what to expect.


The alternative approach, is to get rid of this error and add another function similar to "fn retrieve_raw_node" - something like 'retrieve_raw_node_or_value" instead.

Reply:
Hm, I think this case is pretty clear: today, the contract is that, if you call `retrieve_raw_node` and hit the case where the node fails to decode, that means 

* either a programming error
* or the database being seriously messed up and lying to us about its contents. 

So, when the caller gets storage inconsistent state, it's pretty clear what happens, and what can be done (not much). 


Now, in this PR we add one more variant, and the problem here is that now the call-site has no idea what to do with `NotANode`. It's very fuzzy whether this is a programming error, a catastrophic DB failure, an invalid challenge, or maybe just a normal outcome. 

So, yeah, I would much prefer to add 

```rust
fn retrieve_raw_node_or_value() -> Result<Result<Node, NotANode>, StorageError>
```

this won't infect every existing caller of a `StorageError`-returning function. 


---


Looking more holistically here, it seems it should never be the case that we don't know whether we expect a node or a value? The root of the trie is always a node, and, when we descend, we always know whether we are fetching the root node or a value.

In other words, I think the usage of this error in `print_recursive_internal` is, pedantically speaking, incorrect -- it's only on the first level of recursion where we don't know whether this is  node or a value, all recursive calls must treat this condition as a messed-up database. So, we rather want something like this instead: 


```
pub fn print_recursive(&self, hash: &CryptoHash, max_depth: u32) -> String {
        if let Err() = self.get_raw_trie_node(hash) {
            panic!("{hash} doesn't point to a trie node")   
        }
        self.print_recursive_internal(hash)
}
```

Reply:
thanks for the tips - done.